### ISSUE TITLE
Latency with ASR STT

### ISSUE THREAD

**Issue by endomorphosis** [opea-project/GenAIComps#391]
Whisper web takes 2 seconds to transcribe one second of audio in my browser.

ASR in habana takes 2 seconds to transcribe one second of audio through the API.

Can we change or optimize the whisper model so that the latency is reduced?

**Comment by endomorphosis** [comment]
![image](https://github.com/user-attachments/assets/ef267b61-9406-4aaf-a616-d88eb08d6729)

```
import requests
import json
import time
from pydub import AudioSegment
import io
import base64
import numpy as np

def process_audio(audio_path):
    start_time = time.time()

    # Load and convert audio to 16000 Hz mono
    waveform = AudioSegment.from_file(audio_path).set_frame_rate(16000).set_channels(1)

    # Convert AudioSegment to numpy array
    samples = np.array(waveform.get_array_of_samples())

    # Convert numpy array to bytes
    byte_io = io.BytesIO()
    waveform.export(byte_io, format="wav")
    audio_bytes = byte_io.getvalue()

    # Encode audio bytes to base64
    audio_base64 = base64.b64encode(audio_bytes).decode('utf-8')

    # Prepare the API request
    url = "http://159.69.148.218/v1/audio/transcriptions"
    headers = {'Content-Type': 'application/json'}
    data = {
        "byte_str": audio_base64
    }

    # Send the request to the API
    response = requests.post(url, headers=headers, data=json.dumps(data))

    # Check if the request was successful
    if response.status_code == 200:
        transcript = response.json().get('text', '')
    else:
        transcript = f"Error: {response.status_code} - {response.text}"

    end_time = time.time()
    execution_time = end_time - start_time

    return transcript, execution_time
```

**Comment by endomorphosis** [comment]
```
import requests
import json
import time
from pydub import AudioSegment
import io
import base64
import numpy as np

def process_audio(audio_path):
    start_time = time.time()

    # Load and convert audio to 16000 Hz mono
    waveform = AudioSegment.from_file(audio_path).set_frame_rate(16000).set_channels(1)

    # Convert AudioSegment to numpy array
    samples = np.array(waveform.get_array_of_samples())

    # Convert numpy array to bytes
    byte_io = io.BytesIO()
    waveform.export(byte_io, format="wav")
    audio_bytes = byte_io.getvalue()

    # Encode audio bytes to base64
    audio_base64 = base64.b64encode(audio_bytes).decode('utf-8')

    # Prepare the API request
    url = "http://159.69.148.218/v1/audio/transcriptions"
    headers = {'Content-Type': 'application/json'}
    data = {
        "byte_str": audio_base64
    }

    # Send the request to the API
    response = requests.post(url, headers=headers, data=json.dumps(data))

    # Check if the request was successful
    if response.status_code == 200:
        transcript = response.json().get('text', '')
    else:
        transcript = f"Error: {response.status_code} - {response.text}"

    end_time = time.time()
    execution_time = end_time - start_time
    return transcript, execution_time
```
![image](https://github.com/user-attachments/assets/cd554088-8b2e-4fd2-ad31-b23f3b21aca9)

**Comment by endomorphosis** [comment]
![image](https://github.com/user-attachments/assets/1d18a80e-59c9-4dac-805c-4bc0d67a4273)


I suggest https://huggingface.co/distil-whisper/distil-large-v3

**Comment by endomorphosis** [comment]
![image](https://github.com/user-attachments/assets/c6016d95-e1a0-45a8-ba15-89a6d3bee27c)

The issue is that intel devcloud wont let me stage a kubernetes cluster, so i am using a reverse proxy via digital ocean Seattle, and i am in portland oregon . The reverse proxy is adding some latency overhead, and the user (Christoph Schuhmann) was in Germany, so I think the rest of the latency is a result of his location and his python environment.

![image](https://github.com/user-attachments/assets/682aba7e-fa09-4acd-a9d8-00a40da7fa84)

I tried using distill whisper, but this did not improve the latency with the HPU (presumably because it has enough memory bandwidth to make it not a limiting factor, compared to the CPU/network latency)

**Comment by endomorphosis** [comment]
tested with a one second wav file
Groq
![image](https://github.com/user-attachments/assets/653ae3a4-e295-46f0-a740-53c9bbf7f223)

openai
![image](https://github.com/user-attachments/assets/838273c0-5240-40ed-8d40-7021c300c483)

Cloudflare
![image](https://github.com/user-attachments/assets/29c8277a-032c-4437-b574-ba0d17cfd44d)
tested with bytestring example
### 

Xeon
localhost
![image](https://github.com/user-attachments/assets/86e64aa4-fbba-4e98-bfdb-d90edafacfe9)

reverse proxy devcloud-> seattle (digital ocean) -> intel devcloud -> seattle (digital ocean) -> devcloud

![image](https://github.com/user-attachments/assets/8306d541-f09c-4ddd-b935-f3387b63d980)

HPU
top (reverse proxy) bottom: localhost
![image](https://github.com/user-attachments/assets/e4655128-6ba4-4f58-ba19-b86e067f55a4)

**Comment by endomorphosis** [comment]
Yes, I agree, but i cannot always rely on the the enduser environment being able to run onnx models performantly. 

I have also told him that he should be developing this "classroom buddy" as a javascript frontend with onnx backend and not a python application, but unfortunately the people who are doing most of the development are college students, and I do not pay them so I cannot boss them around.

I can only try to control the chaos, by forcing them to use OPEA, and keeping the UI / application / inference logic separate.
![image](https://github.com/user-attachments/assets/459eb1be-587a-4774-bc4a-4ef4700835b7)

**Comment by endomorphosis** [comment]
![image](https://github.com/user-attachments/assets/a95bcc83-8fbd-46d6-aba1-e472b475b09e)

Christoph reports that even with the provided asr example, that the response takes too long, and was wondering whether it would be possible to do 8xGaudi for the ASR service to make it as fast as  groq (or even openai api)

**Comment by endomorphosis** [comment]
https://github.com/Vaibhavs10/insanely-fast-whisper

This is an example of a optimum (cuda) based batch inference solution

**Comment by endomorphosis** [comment]
Christoph is a part of the Intel Center for Excellence 

https://www.linkedin.com/in/christoph-schuhmann-59a740235/?originalSubdomain=de
https://www.bloomberg.com/news/features/2023-04-24/a-high-school-teacher-s-free-image-database-powers-ai-unicorns

He is working on a project called Bud-E
https://youtu.be/fTpVsN6dUNM

The goal is to get the latency down for larger models and retrieval, but he would like to keep the whisper transcription part to less than 500ms, so as to keep the time to produce the first audio bytes output to below 2 seconds for the entire pipeline including llm and retrieval.

**Comment by endomorphosis** [comment]
I guess this is more of a feature request than a bug fix, the request is that the ASR STT itself supports batch inference, and that the "insanely fast whisper" substitute optimum for optimum habana gaudi, because the batch inference solution the package provides can perform 150 minutes of transcription in 1 min 18 sec, or rather 115x real time speed vs 6x real time speed for the current implementation.

**Comment by endomorphosis** [comment]
> How do you get "6x real time speed"? Can you provide your steps?
I assume your audio is one second and divide 1 / 0.17714595794677734
 
> "150 minutes of transcription in 1 min 18 sec", how or where do you get this data?
I assume that a Gaudi GPU is about as fast an an A100.
https://github.com/Vaibhavs10/insanely-fast-whisper

![image](https://github.com/user-attachments/assets/95bd4249-6b48-40c6-8b02-2a6fb980fbfb)

**Comment by endomorphosis** [comment]
https://x.com/reach_vb/status/1820560137892835369
![image](https://github.com/user-attachments/assets/38687a85-e739-46f7-9caa-81e17c9cd20d)

**Comment by endomorphosis** [comment]
yes, but the number of concurrent users per microservice and the latency could be vastly improved, I understand that I'm not your boss, but this is a feature that I would like.

I am working on getting a feature for llama 405b working, so that I can work on a graphrag microservice

**Comment by endomorphosis** [comment]
I have a company in Oregon that is trying to do legal automation with a virtual avatar, because of an acute lawyer shortage in Oregon, I used to work as a cloud infrastructure engineer at Jones Farm.

I volunteer with LAION.AI (Intel COE) to publish open source 3d animated avatars, and finetune audio tokens into llama3 to improve latency 
I volunteer with libp2p / ipfs / protocol labs / filecoin foundation to create a new open source peer to peer based mlops system from scratch focusing on privacy preserving edge compute. 
I volunteer with the Stanford Codex computational law research paper reading group, and I was trying to have llama 405B fp8 / opea ready to demo for the stanford law x llm hackathon.
I volunteer with the  Yannic Kilcher machine learning discord weekly/daily research paper reading along with Desmond Greely at Intel liftoff.

My cofounder works for Edgerunner.ai, where he has a license to my edge focused mlops project

The other cofounder runs a cryptocurrency exchange, and a video game product.

**Comment by Spycsh** [comment]
Hi @endomorphosis , thanks for raising this issue. The original data you showed is 2 seconds for HPU inference of 1 second audio. Since you found it was because of the network latency, let's proceed to the context when it took ~3 seconds to transcribe the text given by Christopher. This could happen because Habana currently needs a warmup before doing inference. I actually use a 15 seconds input audio to do the [warmup](https://github.com/opea-project/GenAIComps/blob/main/comps/asr/whisper/whisper_model.py#L39) so any input audio that less than 15 seconds (may not be exact but close to the upper bound) can fully get the use of HPU graph.  Maybe Christopher's input audio is larger than 15 seconds so it may be the reason that you see it takes ~3 seconds to do an inference of the given text. One simple check is that you can run multiple times the same input audio and check whether the latency drops a lot.

Simultaneously, I will try to reproduce and figure out a proper way to narrow down the Whisper inference time for HPU for a single instance, and fix any issues that I encounters.

Apart from the HPU issue, the latency of ASR can be improved in other few aspects that we may contribute to:

1. inference parallelism (as you said, multiple requests of different audio lengths can be batched together to do an inference)
2. split long audio into multiple segments and batch the segments to do an inference (as the [repo](https://github.com/Vaibhavs10/insanely-fast-whisper/tree/main) you mentioned shows, correct me if I'm wrong!)
3. update models (whisper-large, whisper-small, whisper-tiny, distil-whisper etc.)
4. continuous batching of Whisper model (utilizing the inference frameworks such as vLLM, TGI, etc.)
5. Low-precision (bf16, fp16, fp8, int8...)
6. Scale out more microservice instances for serving

Regarding to 4,5, I believe them are some study already and will tune to the latest open-source study and integrate it asap. Regarding to 3, I previously also try distil-whisper but I do not see a clear latency drop, so I think you can choose the model that you try mostly fit your demand. Regarding to 1, I believe it is an alternative of 4 so we will measure which way is a better way for Whisper. Regarding to 2, yes for OPEA, we currently mainly consider the Talkingbot daily conversation with LLMs so it assumes user do not need to say a super long speech as input. But would be good to support that. Regarding to 6, yes you can try scaling out multiple ASR instances on multiple Gaudi cards.

I'm currently also doing the the related performance tuning stuff of AudioQnA and try to improve latency and throughput for ASR/TTS. Could you also share your scenario so I can also consider your scenario/request of features into the whole design and implementation?

Tell me if you have any other questions!
